mav_trajectory_generation/src/trajectory.cpp

```diff
@@ -264,4 +264,85 @@ Vertex Trajectory::getGoalVertex(int max_derivative_order) const {
+bool Trajectory::computeMaxVelocityAndAcceleration(double* v_max,
+                                                   double* a_max) const {
+  std::vector<int> dimensions;  // Evaluate in whatever dimensions we have.
+  for (int i = 0; i < D_; i++) {
+    dimensions.push_back(i);
+  }
+  Extremum v_min_traj, v_max_traj, a_min_traj, a_max_traj;
+
+  bool success = computeMinMaxMagnitude(
+      mav_trajectory_generation::derivative_order::VELOCITY, dimensions,
+      &v_min_traj, &v_max_traj);
+  success &= computeMinMaxMagnitude(
+      mav_trajectory_generation::derivative_order::ACCELERATION, dimensions,
+      &a_min_traj, &a_max_traj);
+
+  *v_max = v_max_traj.value;
+  *a_max = a_max_traj.value;
+  return success;
+}
+
+// This method SCALES the segment times evenly to ensure that the trajectory
+// is feasible given the provided v_max and a_max. Does not change the shape
+// of the trajectory, and only *increases* segment times.
+bool Trajectory::scaleSegmentTimesToMeetConstraints(double v_max,
+                                                    double a_max) {
+  double v_max_actual, a_max_actual;
+  computeMaxVelocityAndAcceleration(&v_max_actual, &a_max_actual);
+
+  double velocity_violation = v_max_actual / v_max;
+  double acceleration_violation = a_max_actual / a_max;
+
+  int counter = 0;
+  // In vast majority of cases, this will converge within 1 iteration.
+  constexpr int kMaxCounter = 20;
+  constexpr double kTolerance = 1e-3;
+  bool within_range = velocity_violation <= 1.0 + kTolerance &&
+                      acceleration_violation <= 1.0 + kTolerance;
+
+  while (!within_range && (counter < kMaxCounter)) {
+    // From Liu, Sikang, et al. "Planning Dynamically Feasible Trajectories for
+    // Quadrotors Using Safe Flight Corridors in 3-D Complex Environments." IEEE
+    // Robotics and Automation Letters 2.3 (2017).
+
+    double violation_scaling = std::max(
+        1.0, std::max(velocity_violation, sqrt(acceleration_violation)));
+    std::cout << "Vel violation: " << velocity_violation
+              << " accel violation: " << acceleration_violation
+              << " v max: " << v_max << " a max: " << a_max
+              << " scaling: " << violation_scaling << std::endl;
+
+    // First figure out how to stretch the trajectory in time.
+    double violation_scaling_inverse = 1.0 / violation_scaling;
+
+    // Scale the segment times of each segment.
+    double new_max_time = 0.0;
+    for (size_t i = 0; i < segments_.size(); i++) {
+      double new_time = segments_[i].getTime() * violation_scaling;
+      for (int d = 0; d < segments_[i].D(); d++) {
+        (segments_[i])[d].scalePolynomialInTime(violation_scaling_inverse);
+      }
+      segments_[i].setTime(new_time);
+      new_max_time += new_time;
+    }
+    max_time_ = new_max_time;
+
+    // Reevaluate min/max extrema
+    v_max_actual = 0.0;
+    computeMaxVelocityAndAcceleration(&v_max_actual, &a_max_actual);
```

Comment:
If you move the within range check to the start of the while loop and break on success you can turn it into a for loop and don't need the initial check beforehand.